core/domain/translation_domain.py

```diff
@@ -381,3 +703,497 @@ def to_dict(self) -> Dict[str, str]:
+                    language_code_to_written_translation.keys()):
+                translation_dict = written_translations_dict[
+                    'translations_mapping'][content_id][language_code]
+                if 'data_format' in translation_dict:
+                    if (translation_dict['data_format'] ==
+                            WrittenTranslation.DATA_FORMAT_HTML):
+                        translation = written_translations_dict[
+                                'translations_mapping'
+                            ][content_id][language_code]['translation']
+                        # Ruling out the possibility of any other type for mypy
+                        # type checking.
+                        assert isinstance(translation, str)
+                        written_translations_dict['translations_mapping'][
+                            content_id][language_code]['translation'] = (
+                                conversion_fn(translation)
+                            )
+                elif 'html' in translation_dict:
+                    # TODO(#11950): Delete this once old schema migration
+                    # functions are deleted.
+                    # This "elif" branch is needed because, in states schema
+                    # v33, this function is called but the dict is still in the
+                    # old format (that doesn't have a "data_format" key).
+                    # In convert functions, we allow less strict typing
+                    # because here we are working with previous versions of
+                    # the domain object and in previous versions of the domain
+                    # object there are some fields that are discontinued in
+                    # the latest domain object (eg. html). So, while accessing
+                    # these discontinued fields MyPy throws an error. Thus to
+                    # avoid the error, we used ignore here.
+                    written_translations_dict['translations_mapping'][
+                        content_id][language_code]['html'] = (  # type: ignore[misc]
+                            conversion_fn(translation_dict['html']))  # type: ignore[misc]
+
+        return written_translations_dict
+
+
+class ContentIdGenerator:
+    """Class to generate the content-id for a translatable content based on the
+    next_content_id_index variable.
+    """
+
+    def __init__(self, start_index: int = 0) -> None:
+        """Constructs an ContentIdGenerator object."""
+        self.next_content_id_index = start_index
+
+    def generate(
+        self,
+        content_type: ContentType,
+        extra_prefix: Union[str, None] = None
+    ) -> str:
+        """Generates the new content-id from the next content id."""
+        content_id = content_type.value + '_'
+        if extra_prefix:
+            content_id += extra_prefix + '_'
+        content_id += str(self.next_content_id_index)
+
+        self.next_content_id_index += 1
+        return content_id
```

Comment:
Hmm, I'm a bit confused by this. Shouldn't the `next_content_id_index` be separate for each content type and extra prefix?

Reply:
next_content_id_index is a global (for each exploration) index for creating content_id and it doesn't depend on where it's used, the overall generated id will be unique! 

Note: The prefix (like hint, content etc.) in the content id is just for having a readable content_id.